File: go/src/275.h-指数-ii.go
/*
 * @lc app=leetcode.cn id=275 lang=golang
 *
 * [275] H 指数 II
 */

// @lc code=start
func hIndex(citations []int) int {
	n := len(citations)
	l, r := 0, n
	for l < r {
		mid := l + (r-l)>>1

		if citations[mid] < n-mid {
			l = mid + 1
		} else {
			r = mid
		}
	}

	return n - l
}

// @lc code=end

File: rust/src/275.h-指数-ii.rs
/*
 * @lc app=leetcode.cn id=275 lang=rust
 *
 * [275] H 指数 II
 */

// @lc code=start
impl Solution {
    pub fn h_index(citations: Vec<i32>) -> i32 {
        let n = citations.len() as i32;
        if n == 0 || citations[n as usize - 1] == 0 {
            return 0;
        }

        let (mut left, mut right) = (0, n - 1);

        while left < right {
            let mid = left + (right - left) / 2;
            if citations[mid as usize] < n - mid {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        n - left
    }
}
// @lc code=end
